### ISSUE TITLE
Getting `RawArgs` without advancing the iterator

### ISSUE THREAD

**Issue by tertsdiepraam** [blyxxyz/lexopt#15]
Hi! I'm writing a library on top of `lexopt` and I'd like to peek the next argument without advancing the optional value, because I need to support some weird arguments that don't follow the usual rules.

Essentially I'd like a method like this, which gives me the `RawArgs` if there are no more shorts or long values to take:
```rust
pub fn try_raw_args(&mut self) -> Option<RawArgs<'_>> {
    if self.long_value.is_some() || self.shorts.is_some() {
         return None;
    }

    Some(RawArgs(&mut self.source))
}
```

With this method, I can make a loop like this, where the check for the custom argument does not interfere with the normal parsing:

```rust
loop {
    // Check the weird custom argument
    if let Some(raw) = parser.try_raw_args() {
        ...
    }

    // Parse normal argument
    let Some(arg) = parser.next() else { break; };
    ...
}
```

**Comment by blyxxyz** [comment]
This makes sense to support!

Usually `x()` vs `try_x()` is panicking vs erroring, so the name is a little unidiomatic. But it does sort of fit the nonrecoverable/recoverable pattern, and I can't think of a better one.

*Maybe* a separate method could provide this information, in case you don't need to look at the arguments themselves or can't borrow mutably. But I don't have any specific cases in mind, so this seems like a good shape for the API.

I have an unpushed commit that adds a `RawArgs::as_slice(&self) -> &[OsString]` method, that might interest you as well. (I'll try to push some changes soonish and then work on this.)

**Comment by tertsdiepraam** [comment]
> Usually x() vs try_x() is panicking vs erroring, so the name is a little unidiomatic. But it does sort of fit the nonrecoverable/recoverable pattern, and I can't think of a better one.

Agreed! I couldn't think of a better name either, so this is what I went with, but I'll keep thinking about a better one.

> Maybe a separate method could provide this information, in case you don't need to look at the arguments themselves or can't borrow mutably. But I don't have any specific cases in mind, so this seems like a good shape for the API.

Yes, a method like `has_optional_value` could work as well. Then it would be:
```rust
loop {
    if !parser.has_optional_value() {
        // get raw args always with the guarantee that it doesn't return an error and therefore doesn't advance
    }
   
    // normal parsing
}
```

I don't really mind what the exact API is, you're probably in a better position to decide on that :)

> I'll try to push some changes soonish and then work on this.

Thank you! I'd also be happy to open up a PR once you know what it should look like.

**Comment by blyxxyz** [comment]
I added this in a9c6031c20bfa566bcf1f596f80f969a7cbfb867 in the `develop` branch. Can you try it for your use case?

**Comment by tertsdiepraam** [comment]
Works like a charm! Thank you! The code below is a bit quick and dirty but shows what I was trying to do. Do you want to keep it as `try_raw_args` or are you thinking of other names?

```rust
use std::ffi::OsStr;

use lexopt::prelude::*;

fn main() -> Result<(), lexopt::Error> {
    let mut parser = lexopt::Parser::from_env();
    loop {
        // Check the custom argument consisting of a hyphen followed by digits
        if let Some(number) = parser
            .try_raw_args()
            .as_mut()
            .and_then(|a| a.peek())
            .and_then(OsStr::to_str)
            .and_then(|s| s.strip_prefix('-'))
            .and_then(|s| s.parse::<i64>().ok())
        {
            println!("Got number {number}");
            
            // Skip over the value we just parsed
            parser.value().unwrap();
        }

        if let Some(arg) = parser.next()? {
            match arg {
                Short('f') | Long("foo") => {
                    println!("Got foo!");
                }
                _ => {
                    panic!("Unexpected argument");
                }
            }
        } else {
            break;
        }
    }
    Ok(())
}
```

**Comment by blyxxyz** [comment]
Great! I'll stick with this name unless I suddenly come up with a better one.

I'll also add more documentation, including an example of your use case.

Maybe a `peek_raw(&self) -> Option<&OsStr>` method would be more convenient for this use case, so that's worth thinking about. But that can come in another release, if at all.

**Comment by blyxxyz** [comment]
Hmmm. A few observations after working on an example.

- The second `if let` should be `else if let` to work as intended.
- `peek()` can take `&self` now. I'll change that, this is a breaking release anyway.
- `Parser::value()` works fine but `RawArgs::next()` is more obviously correct.
- A `RawArgs::parse_with<T>(impl FnOnce(&str) -> Option<T>) -> Option<T>` kind of function would be convenient for this. It could take care of consuming the argument for you. But that's not worth rushing into the next release, if it's worth doing at all.
- Things become inconvenient if `RawArgs` is a temporary. Your example needs `as_ref`/`as_mut` and so would [`targo`](https://github.com/sunshowers/targo/blob/7518c97b52ffc84beea42a02f520c58d49696d16/targo/src/dispatch.rs#L163) (which checks for `--` without consuming it).
  - A function that consumes RawArgs or goes directly from `&Parser` to `&OsStr`/`&[OsStr]` wouldn't have this issue.
  - An immutable version of RawArgs wouldn't have this issue.
- (Parsing `i64` means you match double hyphens.)
- (GNU `tail` and `head` only seem to accept this form as the first argument, so you'd perhaps want to do this outside the loop.)

Nothing blocking, I think, so I hope to do a release soon.

My version of the example may look something like this:
```rust
use std::path::PathBuf;

fn parse_dashnum(parser: &mut lexopt::Parser) -> Option<u64> {
    let mut raw = parser.try_raw_args()?;
    let arg = raw.peek()?.to_str()?;
    let num = arg.strip_prefix('-')?.parse::<u64>().ok()?;
    raw.next(); // Consume the argument we just parsed
    Some(num)
}

fn main() -> Result<(), lexopt::Error> {
    use lexopt::prelude::*;

    let mut parser = lexopt::Parser::from_env();
    loop {
        if let Some(num) = parse_dashnum(&mut parser) {
            println!("Got number {}", num);
        } else if let Some(arg) = parser.next()? {
            match arg {
                Short('f') | Long("follow") => {
                    println!("Got --follow");
                }
                Value(path) => {
                    let path = PathBuf::from(path);
                    println!("Got file {}", path.display());
                }
                _ => return Err(arg.unexpected()),
            }
        } else {
            break;
        }
    }

    Ok(())
}
```

**Comment by tertsdiepraam** [comment]
I get free code review😄 I agree with all your points and your version is a clear improvement over mine. Thanks!

**Comment by blyxxyz** [comment]
This is now released in v0.3.0. Thank you for the suggestion!
